DESIGN.md

```diff
@@ -0,0 +1,99 @@
+# Shipit - Opinionated release management for Rust crates
+
+Shipit is intended to aid in the release management process of Rust projects
+including Tokio, Tower, and Mio. While it offers some level of configuration, it
+is opinionated in the process that developers will take. Projects that stick
+with the Shipit process will gain automated publishing to crates.io including
+changelog generation.
+
+## Project workflow with Shipit.
+
+In order to work with Shipit, projects should be contained in a [Git] repository
+and structured to follow [Cargo] idioms. Shipit supports both single crates and
+[Cargo workspaces][workspace] containing multiple crates. When the repository
+contains a workspace with multiple crates, additional requirements are described
+below.
+
+Shipit requires that the master branch is always able to be released. This means
+that changes must be fully implemented in branches and merged in PRs. It also
+means that breaking changes should not be applied to the master branch until the
```

Comment:
Is it the job of the PR that creates the breaking change to also bump the version? Meaning, as soon as that PR is merged there should be a new unreleased version ready?

Reply:
The way I'm thinking about it, the PR will never bump the version. I don't see a reason to bump the version manually if a) the PR is tagged as breaking and b) master is can always be published. The implication is that breaking changes should be queued up in unmerged PRs or some other branch until ready to publish.

Reply:
So if a PR is labeled as breaking, when does the actual version get bumped? Since, the next planned release? or will once that PR gets merged a new one opens from the bot with the ability to release that?

Reply:
What do you think would be best? It does seem like bumping early is correct, especially to get workspace dependencies lined up...

Reply:
I personally think if a PR introduces a breaking change, that PR should be responsible for fixing all the breaking change issues. For example I made this breaking change in `tower-util` and it doesn't compile the rest of the crates till they are all updated. This means, CI is enforcing that this PR doesnt get merged until I have fixed all the remaining apis that would use this. I think this would also make it nice to localize breaking changes to one PR, instead of allowing partial work to be introduced that might block releases.

https://github.com/tower-rs/tower/pull/144/files#diff-51345482bb3da2a7549c5ed440228063R3

Reply:
Ok, I think that is fair. It also works with long lived "breaking change" branches. When the breaking changes are ready to be released, that branch can be merged into master via a PR.